lightningd/plugin.c

```diff
@@ -491,30 +491,41 @@ static struct io_plan *plugin_read_json(struct io_conn *conn,
 					struct plugin *plugin)
 {
 	bool success;
+	bool have_full;
+	int offset = plugin->used >= 1 ? 1 : 0;
 
 	log_io(plugin->log, LOG_IO_IN, NULL, "",
 	       plugin->buffer + plugin->used, plugin->len_read);
 
+	/* Peek into the buffer increment we just read (including a tiny
+	 * overlap if we're not at the start) to see if we have a '\n\n'
+	 * separator. If not we can skip the JSON attempt. */
+	have_full = memmem(plugin->buffer + plugin->used - offset,
+			   plugin->len_read + offset, "\n\n", 2);
+
```

Comment:
Plugins are *not* obliged to do a double CR.

This doesn't work.

Reply:
I also don't see how it would help, since plugin_read_json_one() is not O(n^2) any more?

Reply:
But it is, just look at the last two flamegraphs, with the `jsmn_parse_string` all but disappearing. I expect jsmn continues parsing from the last complete token, but will re-parse any token that was truncated. If we have really large fields (like a 2MB hex-encoded string) it'll parse that several times over and over.

I think we should make `\n\n` mandatory, just for this reason alone.

Reply:
Alternatively we can also match on `}\n` which might trigger a bit more often for nested structures (which as generally smaller) but is guaranteed to terminate a message ("a JSON object terminated by a newline").